DUPLICATE: Function encoding

ana-pantilie opened the pull request.

andreistefanescu commented:
@dwightguth please review

dwightguth commented:
looks good but there's a couple minor improvements I'd like to see with respect to the isGround and variableSet fields. Namely, if isGround is true, we should be able to deduce that variableSet is empty without traversing the entire term, and similarly, if we have already computed the variableSet, we should be able to determine  isGround without traversing the term by checking the size of the set.

Also, do you have any statistics on the effect this diff has on performance? I'm assuming you did it because it was faster but I'd be interested to see numbers before we merge this.

andreistefanescu commented:
The old version is about 50% slower. This does not mean much for concrete execution, but it's useful for verification.

dwightguth commented on java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java:
```diff
@@ -113,39 +110,51 @@ public JavaSymbolicObject substitute(Variable variable, Term term, TermContext c
+     */
+    public boolean isGround() {
+        if (isGround == null) {
+            if (variableSet == null || !variableSet.isEmpty()) {
```
`if (variableSet == null) {`

dwightguth commented on java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java:
```diff
@@ -113,39 +110,51 @@ public JavaSymbolicObject substitute(Variable variable, Term term, TermContext c
+            if (variableSet == null || !variableSet.isEmpty()) {
+                new IsGroundFieldInitializer().visitNode(this);
+            } else {
+                isGround = true;
```
`isGround = variableSet.isEmpty();`

andreistefanescu commented:
On a side note, ```hashCode``` is a lot faster than ```isGround```, so all the boilerplate code is slowing things down. Unfortunately, it's not feasible to replace all the visitor/transformer classes with handcrafted code.

dwightguth commented:
still one minor fix to make, then you can merge.

dwightguth commented:
lgtm. merge if tests pass.

Commit:
rename some references to old syntax

Commit:
new encoding of functions

Commit:
don't add simplification automatically

Commit:
fix non owise rules

Commit:
disallow priority or owise + simplification

Commit:
restore old behavior for simplification rules

Commit:
tmp: update submodule

Commit:
Merge branch 'master' into function

Commit:
update llvm backend submodule

ttuegel committed:
domains.k: Do not combine priority and simplification attributes

The Haskell backend now supports a simplification attribute with optional
priority argument.

ttuegel committed:
Update haskell-backend

ttuegel committed:
Merge branch 'master' into function

ttuegel committed:
Update llvm-backend

ana-pantilie committed:
Merge remote-tracking branch 'upstream/master' into function

ana-pantilie committed:
rat.md: add simplification attribute for ceil rule
